contracts/src/interaction/SwaprSavingsXDaiRouter.sol

```diff
@@ -0,0 +1,362 @@
+
+    function redeem(
+        uint256 shares,
+        address receiver
+    ) external returns (uint256);
+
+    function redeemXDAI(
+        uint256 shares,
+        address receiver
+    ) external payable returns (uint256);
+}
+
+interface IERC4626 {
+    function approve(address guy, uint256 wad) external returns (bool);
+
+    function previewDeposit(uint256 assets) external view returns (uint256);
+
+    function previewRedeem(uint256 shares) external view returns (uint256);
+
+    function balanceOf(address owner) external returns (uint256);
+}
+
+/// @dev Swap Router specialized in swapping tokens with sDAI, xDAI and wxDAI.
+/// In multi hoop swaps (e.g., a swap OUTCOME<>wxDAI that is expanded to OUTCOME<>sDAI<>wxDAi)
+/// it does a single swap OUTCOME<>sDAI and then redeems the sDAI to wxDAI, getting a better price than using the sDAI<>wxDAI pool.
+contract SwaprSavingsXDaiRouter is ISingleSwapRouter, ISingleQuoter {
+    ISingleSwapRouter public constant swaprRouter =
+        ISingleSwapRouter(0xfFB643E73f280B97809A8b41f7232AB401a04ee1); // Swapr Router address
+    ISingleQuoter public constant swaprQuoter =
+        ISingleQuoter(0xcBaD9FDf0D2814659Eb26f600EFDeAF005Eda0F7); // Swapr Quoter address
+
+    address public constant xDAI =
+        address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE); // dummy xDAI address
+    address public constant wxDAI =
+        address(0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d); // wxDAI address
+    IERC4626 public constant sDAI =
+        IERC4626(0xaf204776c7245bF4147c2612BF6e5972Ee483701); // sDAI address
+
+    SavingsXDaiAdapter public constant savingsXDaiAdapter =
+        SavingsXDaiAdapter(0xD499b51fcFc66bd31248ef4b28d656d67E591A94); // SavingsXDaiAdapter address
+
+    function quoteExactInputSingle(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint160 limitSqrtPrice
+    ) public override returns (uint256 amountOut, uint16 fee) {
+        if (tokenIn == xDAI || tokenIn == wxDAI) {
+            // (w)xDAI<>OUTCOME_TOKEN
+            // deposit (w)xDAI to sDAI and quote sDAI<>OUTCOME_TOKEN
+
+            return
+                swaprQuoter.quoteExactInputSingle(
+                    address(sDAI),
+                    tokenOut,
+                    sDAI.previewDeposit(amountIn),
+                    limitSqrtPrice
+                );
+        }
+
+        if (tokenOut == xDAI || tokenOut == wxDAI) {
+            // OUTCOME_TOKEN<>(w)xDAI
+            // quote OUTCOME_TOKEN<>sDAI and redeem sDAI for (w)xDAI
+            (uint256 quotedAmountOut, uint16 quotedFee) = swaprQuoter
+                .quoteExactInputSingle(
+                    tokenIn,
+                    address(sDAI),
+                    amountIn,
+                    limitSqrtPrice
+                );
+            return (sDAI.previewRedeem(quotedAmountOut), quotedFee);
+        }
+
+        revert("Not OUTCOME<>(s|w|wx)DAI swap");
```

Comment:
should be "Not OUTCOME<>(s|x|wx)DAI swap". However, I don't see how we can swap directly from sDAI <>OUTCOME here.